### ISSUE TITLE
Question about implementing custom widgets

### ISSUE THREAD

**Issue by eemed** [hecrj/iced#768]
I'm trying to create a custom widget that I can easily use like this
```rust
let editor_col = Column::new()
    .push(editor)
    .width(Length::Fill)
    .height(Length::Fill);
```

I started implementing the trait according to the example
```rust
impl<Message, B> Widget<Message, Renderer<B>> for Editor
where
    B: Backend,
{
    fn width(&self) -> Length {
        todo!()
    }

    fn height(&self) -> Length {
        todo!()
    }

    fn layout(&self, renderer: &Renderer<B>, limits: &layout::Limits) -> layout::Node {
        todo!()
    }

    fn draw(
        &self,
        renderer: &mut Renderer<B>,
        defaults: &Defaults,
        layout: Layout<'_>,
        cursor_position: Point,
        viewport: &Rectangle,
    ) -> (Primitive, mouse::Interaction) {
        let mut column: Column<Message> = Column::new();


        if let Some(buf) = self.editor.buffer(self.buf_idx) {
            for line in buf.lines() {
                if let Some(line) = line.as_str() {
                    let row = Row::new().push(Line::new(line.to_string()));
                    column.push(row);
                }
            }
        }

        column.draw(
            renderer,
            defaults,
            layout,
            cursor_position,
            viewport
            )
    }

    fn hash_layout(&self, state: &mut Hasher) {
        todo!()
    }
}

impl<'a, Message, B> Into<Element<'a, Message, Renderer<B>>> for Editor
where
    B: Backend,
{
    fn into(self) -> Element<'a, Message, Renderer<B>> {
        Element::new(self)
    }
}

```

But the renderer parameter is complaining about 

expected mutable reference `&mut iced_graphics::Renderer<iced_wgpu::backend::Backend>`                                                                 
found mutable reference `&mut iced_graphics::Renderer<B>`

**Comment by eemed** [comment]
I would like editor to be a wrapper for rows of text. So it currently just tries to display multiple rows of text.

**Comment by rosenjcb** [comment]
I frankensteined an example and managed to get it working. Try using the `iced_native::Column` instead of the `iced::Column`

```rust
mod circle { 
    use iced_graphics::{Backend, Defaults, Primitive, Renderer, Column};
    use iced_native::{
        layout, mouse, Background, Color, Element, Hasher, Layout, Length,
        Point, Rectangle, Size, Widget,
    };

    pub struct Circle {
        radius: f32,
    }

    impl Circle {
        pub fn new(radius: f32) -> Self {
            Self { radius }
        }
    }

    impl<Message, B> Widget<Message, Renderer<B>> for Circle
    where
        B: Backend,
    {
        fn width(&self) -> Length {
            Length::Shrink
        }

        fn height(&self) -> Length {
            Length::Shrink
        }

        fn layout(
            &self,
            _renderer: &Renderer<B>,
            _limits: &layout::Limits,
        ) -> layout::Node {
            layout::Node::new(Size::new(self.radius * 2.0, self.radius * 2.0))
        }

        fn hash_layout(&self, state: &mut Hasher) {
            use std::hash::Hash;

            self.radius.to_bits().hash(state);
        }

        fn draw(
            &self,
            _renderer: &mut Renderer<B>,
            _defaults: &Defaults,
            layout: Layout<'_>,
            _cursor_position: Point,
            _viewport: &Rectangle,
        ) -> (Primitive, mouse::Interaction) {
            let mut column: Column<Message, B> = Column::new();

            column.draw(
                _renderer,
                _defaults,
                layout,
                _cursor_position,
                _viewport
                )
            }
    }


    impl<'a, Message, B> Into<Element<'a, Message, Renderer<B>>> for Circle
    where
        B: Backend,
    {
        fn into(self) -> Element<'a, Message, Renderer<B>> {
            Element::new(self)
        }
    }
}
```

**Comment by eemed** [comment]
Thanks I got it working using the `iced_native::widget::Column` in the Sandbox trait implementation and adapting my custom widget according to that circle widget implementation.

**Comment by rosenjcb** [comment]
This still seems very weird though. I don't know why `&mut iced_graphics::Renderer<iced_wgpu::backend::Backend>` isn't a valid `Backend` implementation. Ideally, the widgets API should be platform agnostic and the implementation details should be hidden. I.e. the concept of a `Column` or `Container` or `Text` or anything should have their own default implementations in the `iced_wgpu`, `iced_glow` or whatever other renderer is. 

I tried getting around your problem by calling `iced_graphics` wherever I can, but then something else broke. Here's some context:
![image](https://user-images.githubusercontent.com/8102129/111056593-12e40a80-8435-11eb-9008-9b0fdfb5dd11.png)

Basically, there's no way to push a `Text` into this `Column` or `Row` container. I think this has to be a bug. Rust's compiler error is hinting at a lack of a trait implementation, but I'm not sure about that. @hecrj thoughts?
